src/smc-webapp/editor_chat.cjsx

```diff
@@ -699,11 +702,24 @@ ChatRoom = (name) -> rclass
             e.preventDefault()
             @clear_input()
         else if e.keyCode==13 and e.shiftKey # 13: enter key
-            @send_chat(e)
+            @setState(realtime_typing_message_id : '')
+            if not @state['realtime_typing_message_id']
+                @send_chat(e)
+            else
+                @clear_input()
         else if e.keyCode==38 and @refs.input.getValue() == ''
             # Up arrow on an empty input
             @props.actions.set_to_last_input()
 
+    keyup : (e) ->
+        if @state['realtime_typing']
+            if @state['realtime_typing_message_id']
+                @props.actions.send_edit(@props.messages.get(@state['realtime_typing_message_id']), @refs.input.getValue())
+            else
+                mesg = @refs.input.getValue()
+                if mesg.length? and mesg.trim().length >= 1
+                    @setState(realtime_typing_message_id:String(@props.actions.send_chat(mesg)))
```

Comment:
An alternative is to make history different and implement actual real-time editing of chat messages.